docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

```diff
@@ -0,0 +1,1341 @@
+definition, no interface definition, and only
+non-`serializable <https://github.com/digital-asset/daml/blob/main-2.x/sdk/daml-lf/spec/daml-lf-1.rst#serializable-types>`__
+data type definitions. A utility package is typically only made of
+helper functions and constants.
+
+In the following, packages whose LF version does not support upgrades
+(1.15 and earlier) and utility packages are ignored. Everything is as if
+they didn’t exist for the purposes of the package validity check.
+
+Upon upload to a participant, a DAR is checked against previously
+uploaded DARs for upgrade validity. More precisely, for every package
+with name *p* and version *v* present in the uploaded dar:
+
+1. The participant looks up versions *v_prev* and *v_next* of *p* in its package database, such that *v_prev* is the greatest version of
+   *p* smaller than *v*, and *v_next* is the smallest version of *p*
+   greater than *v.*
+
+   .. image:: ./images/upgrade-static-checks-order.svg
+
+2. The participant checks that version *v* of *p* is a valid upgrade of
+   version *v_prev* of p.
+
+3. The participant checks that version *v_next* of *p* is a valid
+   upgrade of version *v* of *p*.
+
+| This means that “being a valid upgrade” for a package is context
+  dependent: it depends on what packages are already installed on the
+  participant.
+| It is also *modular*: checks are performed at the package level. That
+  is, a new version of a package is rejected as soon as it contains some
+  element which doesn’t properly upgrade its counterpart in the old
+  package, even if some other elements do.
+
+Modules
+~~~~~~~
+
+The modules of the upgrading package must form a superset of the modules
+of the upgraded package. In other words, it is valid to add new modules
+but deleting a module leads to a validation error.
+
+Examples
+^^^^^^^^
+
+In the file tree below, package v2 is a potentially valid upgrade of
+package v1, assuming ``v2/A.daml`` is a valid upgrade of ``v1/A.daml``.
+
+.. code::
+
+  ├── v1
+  │   ├── daml
+  │   │   └── A.daml
+  │   └── daml.yaml
+  └── v2
+      ├── daml
+      │   ├── A.daml
+      │   └── B.daml
+      └── daml.yaml
+
+
+In the file tree below, package v2 cannot possibly be a valid upgrade of
+v1 because it doesn’t define module ``B``.
+
+.. code::
+
+  ├── v1
+  │   ├── daml
+  │   │   ├── A.daml
+  │   │   └── B.daml
+  │   └── daml.yaml
+  └── v2
+      ├── daml
+      │   └── A.daml
+      └── daml.yaml
+
+Templates
+~~~~~~~~~~
+
+The templates of the upgrading package must form a superset of the
+templates of the upgraded package. In other words, it is valid to add
+new templates but deleting a template leads to a validation error.
+
+.. _examples-1:
+
+Examples
+^^^^^^^^
+
+Below, the module on the right is a valid upgrade of the module on the
+left. But the module on the left is **not** a valid upgrade of the
+module on the right because it lacks a definition for template ``T2``.
+
+.. list-table::
```

Comment:
Note to the docs team - we would very much like to be able to show code snippets side by side, which would usually work using the syntax below, and `.. code::` blocks on each element.
However, the formatting that docs.daml.com uses for codeblocks forces them to expand the whole width of the screen.
Is there any way we could change this behavior for the blocks in this document? Otherwise, either they don't have code highlighting, or the doc gets twice as long.

Reply:
@paulbrauner-da is also taking a look at this

Reply:
@samuel-williams-da and I finally got to a point where it's reasonably looking. If we can find the time we'll look into improving the margins and getting rid of the drop shadows later.